### PR TITLE
40 fix empty patient search

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes patient related issues in: https://github.com/msupply-foundation/open-msupply-internal/issues/40

# 👩🏻‍💻 What does this PR do? 
- Disable patient search without filter parameters (takes too long with many patients)
- Also limit patient search to 100 on the server
- Fix patient search for prescriptions

# 🧪 How has/should this change been tested? 
~Try to reproduce patient related issues from:
https://docs.google.com/document/d/1UchQFRQwgzdLtsB94uid2XZ-2icsAf6uVNeI1U2vZ5o/edit~

This document has changed too much... Just test that working with the datafile on Android doesn't causes performance issues.

Create new patient Modal:
- Should load reasonably quick
- Clicking generate id should be quick

Which branch should this be merged into? is this a hotfix?

### CODE REVIEW

File: client/packages/common/src/intl/locales/en/dispensary.json [hunk 1]
```diff
@@ -85,7 +85,7 @@
   "messages.error-saving-prescription": "Error saving prescription 🥺",
   "messages.recorded-on": "Recorded on {{datetime}}",
   "messages.results-found": "{{totalCount}} results found",
-  "messages.results-over-limit": "Showing the first 100 results only - please refine your search",
+  "messages.results-over-limit": "Showing the first {{limit}} results only - please refine your search",
```
Comment (clemens-msupply) [inline client/packages/common/src/intl/locales/en/dispensary.json:88]: Thought its nicer to use the limit as returned by the server...

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx [hunk 2]
```diff
@@ -39,18 +39,16 @@ export const SearchWithUserSource = (
...
   const { results, error: queryError, mutateAsync } = usePatientSearchQuery();
 
   useEffect(() => {
+    const searchFilter = !isPatientSelected
+      ? createSearchFilter(options?.searchFields, data)
+      : undefined;
     if (!searchFilter) return;
     mutateAsync(searchFilter);
-  }, [data]);
+  }, [data, isPatientSelected, mutateAsync, options?.searchFields]);
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:51]: Just fixing some warnings

File: client/packages/programs/src/JsonForms/components/Search/usePatientSearchQuery.tsx [hunk 3]
```diff
@@ -1,8 +1,8 @@
-import { Utils } from '@openmsupply-client/system/src/Patient/api/hooks/utils';
 import { PatientSchema } from '@openmsupply-client/programs';
+import { usePatient } from '@openmsupply-client/system';
 
 export const usePatientSearchQuery = () => {
-  const { data, error, isLoading, mutateAsync } = Utils.usePatientSearch();
+  const { data, error, isLoading, mutateAsync } = usePatient.utils.search();
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/usePatientSearchQuery.tsx:5]: Its IMHO still a bit messy but I tried to unify which search function is actually used...

File: client/packages/system/src/Patient/Components/utils.ts [hunk 4]
```diff
@@ -1,27 +1,29 @@
...
+const shouldSearch = (text: string) => text.length > 0;
 export const useSearchPatient = () => {
   const [searchText, setSearchText] = useState('');
   const { mutate, isLoading, data, isSuccess } = usePatient.utils.search();
 
-  const debounced = useDebouncedValueCallback(
-    value => mutate({ nameOrCode: value }),
-    [searchText],
-    500
-  );
```
Comment (clemens-msupply) [inline client/packages/system/src/Patient/Components/utils.ts:13]: Not sure what caused it but for some reason it didn't had the current value of searchText when calling `mutate()`...

File: client/packages/system/src/Patient/Components/utils.ts [hunk 5]
```diff
@@ -1,27 +1,29 @@
...
+  let patients: ProgramPatientRowFragment[] = [];
+  let totalCount = 0;
 
-  useEffect(() => {
-    debounced(searchText);
-  }, [searchText]);
+  if (shouldSearch(debouncedSearchText) && data) {
+    patients = data?.nodes.map(node => node.patient) ?? [];
+    totalCount = data?.totalCount ?? 0;
+  }
```
Comment (clemens-msupply) [inline client/packages/system/src/Patient/Components/utils.ts:26]: This also resets the search when clearing the search input

File: client/packages/system/src/Patient/CreatePatientModal/PatientResultsTab.tsx [hunk 6]
```diff
@@ -154,8 +159,9 @@ export const PatientResultsTab: FC<PatientPanel & { active: boolean }> = ({
   const count = data?.length ?? 0;
 
   useEffect(() => {
+    if (Object.values(searchParams).every(it => it === undefined)) return;
     search(searchParams);
-  }, [patient]);
+  }, [search, searchParams]);
```
Comment (clemens-msupply) [inline client/packages/system/src/Patient/CreatePatientModal/PatientResultsTab.tsx:164]: Don't like how we need to do this "are params empty" check in multiple places but had to fight too many Android build/debug issues that I just did this quick fix for now...

**Review by mark-prins — APPROVED** [review]
👍  all of the reported issues are now fine - the UI is snappy for me on my tablet!

Given the customer decision to limit patient numbers, this is fine to merge into develop rather than hotfix it 🙏

**Comment by clemens-msupply** [comment]
Thanks for the review!

**Pull request merged**